dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -0,0 +1,141 @@
+﻿#include "pch.h"
+#include "PushNotificationManager.h"
+#include "PushNotificationManager.g.cpp"
+#include "PushNotificationCreateChannelResult.h"
+#include <winrt\windows.networking.pushnotifications.h>
+#include <winerror.h>
+#include <algorithm>
+#include <vector>
+#include "PushNotificationChannel.h"
+#include <wil/resource.h>
+
+using namespace winrt::Windows::Networking::PushNotifications;
+using namespace winrt;
+
+namespace winrt::Microsoft::ProjectReunion::implementation
+{
+    inline constexpr std::uint32_t c_maxBackoffSeconds{ 960 };
+
+    const HRESULT WNP_E_NOT_CONNECTED = MAKE_HRESULT(SEVERITY_ERROR, FACILITY_ITF, 0x880403E8L);
+    const HRESULT WNP_E_RECONNECTING = MAKE_HRESULT(SEVERITY_ERROR, FACILITY_ITF, 0x880403E9L);
+    const HRESULT WNP_E_BIND_USER_BUSY = MAKE_HRESULT(SEVERITY_ERROR, FACILITY_ITF, 0x880403FEL);
+
+    bool PushNotificationManager::isChannelRequestRetryable(const hresult& hr)
+    {
+        switch (hr)
+        {
+        case HRESULT_FROM_WIN32(ERROR_TIMEOUT):
+        case WNP_E_NOT_CONNECTED:
+        case WPN_E_OUTSTANDING_CHANNEL_REQUEST:
+        case WNP_E_RECONNECTING:
+        case WNP_E_BIND_USER_BUSY:
+        case RPC_S_SERVER_UNAVAILABLE:
+            return true;
+        default:
+            return false;
+        }
+    }
+
+    inline bool IsPackagedProcess()
+    {
+        uint32_t packageFullNameLength = 0;
+        const auto rc{ GetCurrentPackageFullName(&packageFullNameLength, nullptr) };
+
+        if (rc != ERROR_INSUFFICIENT_BUFFER)
+        {
+            winrt::throw_hresult(rc);
```

Comment:
This doesnt look right. The method says it returns true or false. But here it throws an exception. Are we saying that we get InsufficientBuffer if the process is Packaged? I'm thinking this API needs to be changed.

Reply:
The API definition should be as follows: returns true if packaged, else false. Only throw for critical errors.

Reply:
GetCurrentPackageFullName(&packageFullNameLength, nullptr) - the API here takes a nullptr as the buffer parameter. So we see ERROR_INSUFFICIENT_BUFFER it means that the process is packaged but we couldn't place the packagefullname as there is no buffer pointer passed to it. For all other errors where we either couldn't validate it to be a packaged process or we have seen infra problems so we throw the hresult and return the control back. Thanks.

Reply:
I'm not sure how this is possible. Then the check below would always be true because this API would throw when the app isn't packaged.

Reply:
I just read the [doc about GetCurrentPackageFullName()](https://docs.microsoft.com/en-us/windows/win32/api/appmodel/nf-appmodel-getcurrentpackagefullname). The idea of this logic is fine, I'd just change the last return statement to simply `return true`.

Reply:
Based on the discussion with the team, I have addressed it this way:

IsPackagedProcess() API will:
1.  return true is packaged,
2.  return false if unpackaged - caller of the above API will throw if needed
3.  will throw for all other critical errors.

Thanks.